lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp

```diff
@@ -0,0 +1,137 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Conversion/TTNNToEmitC/TTNNToEmitC.h"
+
+#include "../PassDetail.h"
+#include "PopulatePatterns.h"
+#include "TypeConverter.h"
+
+#include "ttmlir/Dialect/TT/IR/TTOpsDialect.h.inc"
+#include "ttmlir/Dialect/TTNN/IR/TTNN.h"
+#include "ttmlir/Dialect/TTNN/IR/TTNNOps.h"
+
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Func/Transforms/FuncConversions.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+using namespace mlir::tt;
+
+namespace {
+
+void populateTTNNToEmitCPatterns(mlir::MLIRContext *ctx,
+                                 mlir::RewritePatternSet &patterns,
+                                 TypeConverter &typeConverter) {
+  // Device ops
+  //
+  patterns.add<DefaultOpConversionPattern<ttnn::OpenDeviceOp>>(typeConverter,
+                                                               ctx, true);
+  patterns.add<DefaultOpConversionPattern<ttnn::CloseDeviceOp>>(typeConverter,
+                                                                ctx);
+
+  // Memory ops
+  //
+  patterns.add<DefaultOpConversionPattern<ttnn::ToMemoryConfigOp>>(
+      typeConverter, ctx);
+
+  // Tensor ops
+  //
+  patterns.add<DefaultOpConversionPattern<ttnn::FullOp>>(typeConverter, ctx);
+
+  // Math ops
+  //
+  patterns.add<DefaultOpConversionPattern<ttnn::AddOp>>(typeConverter, ctx);
+  patterns.add<DefaultOpConversionPattern<ttnn::SubtractOp>>(typeConverter,
+                                                             ctx);
+  patterns.add<DefaultOpConversionPattern<ttnn::SumOp>>(typeConverter, ctx);
+  patterns.add<DefaultOpConversionPattern<ttnn::MultiplyOp>>(typeConverter,
+                                                             ctx);
+  patterns.add<DefaultOpConversionPattern<ttnn::MatmulOp>>(typeConverter, ctx);
+  patterns.add<DefaultOpConversionPattern<ttnn::ReluOp>>(typeConverter, ctx);
```

Comment:
Each time OP is added it needs to be handled? Will it assert for unknown ops?

Reply:
Correct. I don't like it, but I haven't seen a different way of doing it across several repos. We could do macros, but I don't like that solution... Any ideas?

It won't assert per-se, but the pass will result in an error. For each target (end result of the pass) in dialect conversion API, we can define what dialects and/or ops are legal and illegal - since TTNN dialect is marked as illegal, no ops from that dialect can stay "alive" in the new target, so it will result in an error.

However, if there's no tests to produce these errors, then that's a pickle. There must be some good practices here, I can ask on today's meeting.

Reply:
I am not sure if it would work, but it might be possible to have a base class `TTNNOp` which all of the ttnn ops inherit from, and then have have the following:
```c++
patterns.add<DefaultOpConversionPattern<ttnn::TTNNOp>>(typeConverter, ctx);
```

And possibly override specific ops if different behaviour is needed. So far, the `DefaultOpConversionPattern` uses fields/methods which are common to every op.

Just throwing out ideas, haven't looked enough at how MLIR does things to be sure... 😄

Reply:
I agree, let's leave it as is for this PR and keep looking for other examples that might do something fancier. Definitely no macros, I'd rather copy paste boilerplate.